[AArch64] Change IssueWidth to 6 in AArch64SchedNeoverseV2.td

david-arm opened the pull request:
I think that the issue width for neoverse-v2 CPUs is set too
high and does not properly reflect the dispatch constraints.
I tested various values of IssueWidth (16, 8 and 6) with runs
of SPEC2017 on a neoverse-v2 machine and I got the highest
overall geomean score with an issue width of 6, although it's
only a marginal 0.14% improvement. I also observed a 1-2%
improvement when testing the Gromacs application with some
workloads. Here are some notable changes in SPEC2017 ref
runtimes, i.e. has a ~0.5% change or greater ('-' means
faster):

548.exchange2: -1.7%
510.parest: -0.78%
538.imagick: -0.73%
500.perlbench: -0.57%
525.x264: -0.55%
507.cactuBSSN: -0.5%
520.omnetpp: -0.48%
511.povray: +0.57%
544.nab: +0.65%
503.bwaves: +0.68%
526.blender: +0.75%

If this patch causes any major regressions post-commit it can
be easily reverted, but I think it should be an overall
improvement.

david-arm commented on llvm/test/CodeGen/AArch64/aarch64-reassociate-accumulators-sve.ll:
```diff
@@ -1,4 +1,5 @@
-; RUN: opt -passes=loop-unroll %s -o - | llc -O3 - -mtriple=aarch64-unknown-unknown -mcpu=neoverse-v2 -o - | FileCheck %s
+; RUN: opt -passes=loop-unroll %s -o - | llc -O3 - -mtriple=aarch64-unknown-unknown \
+; RUN:   -mcpu=neoverse-v1 -mattr=+sve2 -o - | FileCheck %s
```
I changed the RUN line here to match the spirit of the test, which is to reassociate accumulators. It was previously relying upon the scheduling of a neoverse-v2 CPU, but I can get the same effect by tuning for neoverse-v1 and enabling SVE2.

david-arm replied:
If I leave the RUN line alone, then the test changes in such a way that it's no longer testing "reassociate chains of accumulation instructions into a tree" (https://github.com/llvm/llvm-project/pull/132728)

sjoerdmeijer commented:
Hi David, I think this makes sense. I would like to run some more benchmarks, and will report back here in a day or two.

david-arm committed:
[AArch64] Change IssueWidth to 6 in AArch64SchedNeoverseV2.td

I think that the issue width for neoverse-v2 CPUs is set too
high and does not properly reflect the dispatch constraints.
I tested various values of IssueWidth (16, 8 and 6) with runs
of SPEC2017 on a neoverse-v2 machine and I got the highest
overall geomean score with an issue width of 6, although it's
only a marginal 0.14% improvement. I also observed a 1-2%
improvement when testing the Gromacs application with some
workloads. Here are some notable changes in SPEC2017 ref
runtimes, i.e. has a ~0.5% change or greater ('-' means
faster):

548.exchange2: -1.7%
510.parest: -0.78%
538.imagick: -0.73%
500.perlbench: -0.57%
525.x264: -0.55%
507.cactuBSSN: -0.5%
520.omnetpp: -0.48%
511.povray: +0.57%
544.nab: +0.65%
503.bwaves: +0.68%
526.blender: +0.75%

If this patch causes any major regressions post-commit it can
be easily reverted, but I think it should be an overall
improvement.

c-rhodes commented:
related #136374

paulwalker-arm reviewed (approved):
If @sjoerdmeijer's data reflects our own measurements then this PR looks go to me.

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:
```diff
@@ -14,7 +14,7 @@
 
 def NeoverseV2Model : SchedMachineModel {
-  let IssueWidth            =  16; // Micro-ops dispatched at a time.
+  let IssueWidth            =   6; // Micro-ops dispatched at a time.
```
Nit: can we update the comment, and try to be more specific what we mean here, because things are all a bit vague in the SWOG and here too. So a bit more rationale that we are now more modelling the dispatch constraints rather than the issue width, and that 6 is better than 8 because that's what benchmarks shows. 
If we get it right here, we might get it right elsewhere too, because these things tend to be copied from one scheduling file to the next one. ;-)

david-arm replied:
I've tried to update the comment, hope it makes sense!

davemgreen replied:
Perhaps just call it the decode bandwidth as that is where the value comes from, and mention that from empirical measurements lower values worked better than higher values?

sjoerdmeijer reviewed (commented):
I suspect we are looking at very similar things. I have tested the interaction between this and #140897 that I plan to commit end of this week. That all looks fine for my motivating examples and app, so that is a good start.
I would like to do some more extensive testing now, and need a bit more time, maybe a day or so.

In the meantime, I've added one nit.

sjoerdmeijer reviewed (approved):
LGTM too, thanks.

davemgreen reviewed (approved):
Cheers
